libs/postgres_ffi/wal_generate/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+        }
+        bail!("Connection timed out");
+    }
+
+    pub fn kill(&mut self) {
+        self.process.kill().unwrap();
+        self.process.wait().unwrap();
+    }
+}
+
+impl Drop for PostgresServer {
+    fn drop(&mut self) {
+        use std::result::Result::Ok;
+        match self.process.try_wait() {
+            Ok(Some(_)) => return,
+            Ok(None) => {
+                warn!("Server was not terminated, will be killed");
+            }
+            Err(e) => {
+                error!("Unable to get status of the server: {}, will be killed", e);
+            }
+        }
+        let _ = self.process.kill();
+    }
+}
+
+pub trait PostgresClientExt: postgres::GenericClient {
+    fn pg_current_wal_insert_lsn(&mut self) -> Result<PgLsn> {
+        Ok(self
+            .query_one("SELECT pg_current_wal_insert_lsn()", &[])?
+            .get(0))
+    }
+    fn pg_current_wal_flush_lsn(&mut self) -> Result<PgLsn> {
+        Ok(self
+            .query_one("SELECT pg_current_wal_flush_lsn()", &[])?
+            .get(0))
+    }
+}
+
+impl<C: postgres::GenericClient> PostgresClientExt for C {}
+
+pub fn generate_simple(client: &mut impl postgres::GenericClient) -> Result<PgLsn> {
+    let initial_lsn = client.pg_current_wal_insert_lsn()?;
+    info!("LSN initial = {}", initial_lsn);
+    client.execute("CREATE table t(x int)", &[])?;
+    let last_lsn = client.pg_current_wal_insert_lsn()?;
+    ensure!(
+        last_lsn == client.pg_current_wal_flush_lsn()?,
+        "WAL is either not flushed or is extended with something unknown"
+    );
+    Ok(last_lsn)
+}
+
+pub fn generate_wal_record_crossing_segment_followed_by_small_one(
+    client: &mut impl postgres::GenericClient,
+) -> Result<PgLsn> {
+    let initial_lsn = client.pg_current_wal_insert_lsn()?;
+    info!("LSN initial = {}", initial_lsn);
+    ensure!(
+        initial_lsn < PgLsn::from(0x0200_0000 - 4 * 8192),
+        "Initial LSN is too far in the future"
+    );
+
+    // This message will be followed by a small COMMIT message and flushed for sure.
+    // Non-transactional messages may be not flushed immediately and wait for the background thread (which flushes every `wal_writer_delay` ms) or following WAL.
+    let message_lsn: PgLsn = client.query_one(
+        "select pg_logical_emit_message(true, 'big-17mb-msg', concat(repeat('abcd', 17 * 256 * 1024), 'end')) as message_lsn",
+        &[]
+    )?.get("message_lsn");
+    ensure!(
+        message_lsn > PgLsn::from(0x0200_0000 + 4 * 8192),
+        "Logical message did not cross the segment boundary"
+    );
+
+    let after_message_lsn = client.pg_current_wal_insert_lsn()?;
+    ensure!(
+        message_lsn < after_message_lsn,
+        "No record found after the emitted message"
+    );
+
+    ensure!(
+        after_message_lsn == client.pg_current_wal_flush_lsn()?,
+        "WAL is either not flushed or is extended with something unknown"
+    );
+    Ok(after_message_lsn)
```

Comment:
And where is small message following crossing record one?

Reply:
It's the implicit COMMIT record.